README.md
angular-vs-repeat v2.0.9
=================

Looking for a version for Angular 2?
===

Check out https://github.com/kamilkp/ng2-vs-for

---

### Breaking changes in v2.x
  * API changes for passing options. Options are now passed as an object to the `vs-repeat` attribute
    instead of using separate attributes for each option (except for `vs-repeat-container`)

### Major changes in v1.0.0-beta.0
  * elements are no longer absolutely positioned
  * no need for a mousewheel helper dummy element on MacOS (to preserve inertia scrolling)
  * no need for `top` updating directives - even better performance

[![Build Status](https://travis-ci.org/kamilkp/angular-vs-repeat.svg?branch=master)](https://travis-ci.org/kamilkp/angular-vs-repeat) [![NPM version](https://badge.fury.io/js/angular-vs-repeat.svg)](http://badge.fury.io/js/angular-vs-repeat) [![Bower version](https://badge.fury.io/bo/angular-vs-repeat.svg)](http://badge.fury.io/bo/angular-vs-repeat)

Virtual Scroll for AngularJS ngRepeat directive

Demo: http://kamilkp.github.io/angular-vs-repeat/

You can find the source code for this demo on branch "gh-pages".

Changelog: https://github.com/kamilkp/angular-vs-repeat/blob/master/CHANGELOG.md

### DESCRIPTION:
`vsRepeat` directive stands for **Virtual Scroll Repeat**. It turns a standard ngRepeated set of elements in a scrollable container
into a component, where the user thinks he has all the elements rendered and all he needs to do is scroll (without any kind of
pagination - which most users loath) and at the same time the browser isn't overloaded by that many elements/angular bindings etc.
The directive renders only so many elements that can fit into current container's `clientHeight`/`clientWidth`.

### LIMITATIONS:
- current version only supports an Array as a right-hand-side object for `ngRepeat`
- all rendered elements must have the same height/width or the sizes of the elements must be known up front

### USAGE:
First include `vs-repeat` as a module dependency in your app.
In order to use the vsRepeat directive you need to place a vs-repeat attribute on a direct parent of an element with ng-repeat
example:

```html
<div vs-repeat="options">
	<div ng-repeat="item in someArray">
		<!-- content -->
	</div>
</div>
```

or:

```html
<div vs-repeat="options">
     <div ng-repeat-start="item in someArray">
         <!-- content -->
     </div>
     <div>
        <!-- something in the middle -->
     </div>
     <div ng-repeat-end>
         <!-- content -->
     </div>
</div>
```

You can also measure the single element's height/width (including all paddings and margins), and then speficy it as a value
of the options' `size` property. This can be used if one wants to override the automatically computed element size.
example:

```html
<div vs-repeat="{size: 50}">	<!-- the specified element height is 50px -->
	<div ng-repeat="item in someArray">
		<!-- content -->
	</div>
</div>
```

- the vsRepeat directive must be applied to a direct parent of an element with `ngRepeat`
- the library also supports ng-repeat-start/ng-repeat-end syntax
- the value of vsRepeat attribute is the single element's height/width measured in pixels. If none provided, the directive will compute it automatically

### OPTIONAL PARAMETERS (attributes):
- vs-repeat-container="selector" - selector for element containing ng-repeat. (defaults to the current element)

### OPTIONS:

Options shall be passed as an object to the `vs-repeat` attribute e.g.:

```html
<div vs-repeat="{scrollParent: 'window', size: 20}"></div>
```

**Available options**:
- `horizontal` - boolean; stack repeated elements horizontally instead of vertically (defaults to false)
- `offset-before` - top/left offset in pixels (defaults to 0)
- `offset-after` - bottom/right offset in pixels (defaults to 0)
- `scroll-margin` - how many pixels ahead should elements be rendered while scrolling (defaults to 0)
- `latch` - if true, elements will be rendered gradually but won't be removed when scrolled away (defaults to false)

- `size` - an angular expression evaluating to the element's size (in pixels) - **it is possible to use the ngRepeat's local repeaing variable in this expression**
- `autoresize` - use this attribute without vs-size and without specifying element's size. The automatically computed element style will
              readjust upon window resize if the size is dependable on the viewport size
- `scrolled-to-end` - callback will be called when the last item of the list is rendered
- `scrolled-to-end-offset` - set this number to trigger the scrolledToEnd callback n items before the last gets rendered
- `scrolled-to-beginning` - callback will be called when the first item of the list is rendered
- `scrolled-to-beginning-offset` - set this number to trigger the scrolledToBeginning callback n items before the first gets rendered

### EVENTS:
- `vsRepeatTrigger` - an event the directive listens for to manually trigger reinitialization
- `vsRepeatResize` - an event the directive listens for to manually trigger the autosizing algorithm
- `vsRepeatReinitialized` - an event the directive emits upon reinitialization done; the listener may accepts three arguments: `event`, `startIndex` and `endIndex`


CHANGELOG.md
Version 2.0.9 (2018/04/02)
=================

  * support for with flex layouts

Version 2.0.7 (2018/03/20)
=================

  * fixed bower.json and added license fields to package.json

Version 2.0.3 (2018/03/16)
=================

  * BREAKING CHANGE: options are now passed as an object to the `vs-repeat` attribute
  * binary search algorithm to find which portion or the original collection should be rendered
  * scope variables like `startIndex`, `endIndex` etc. are now stored under `vsRepeat` object

Version 1.1.11 (2018/03/09)
=================

  * `vs-scroll-margin` option (documentation on that will come shortly)

Version 1.1.10 (2017/03/07)
=================

  * added some logging when there is size mismatch

Version 1.1.7 (2016/03/08)
=================

  * memory leak fixed when using `vs-scroll-parent`

Version 1.1.6 (2016/03/01)
=================

  * table fill content element fix

Version 1.1.5 (2016/02/27)
=================

  * bugfixes and PRs merged
  * use `vs-repeat-container` to specify another repeat conainer (see readme)
  * use `vs-scrolled-to-end` to do something when last element gets rendered (see readme)

Version 1.1.4 (2016/01/17)
=================

  * bugfixes

Version 1.1.2 (2015/12/06)
=================

  * bugfix

Version 1.1.0 (2015/12/04)
=================

  * **breaking change** - the funcitonality implemented in https://github.com/kamilkp/angular-vs-repeat/pull/32 has been removed. It appears it can be implemented independently of vs-repeat and was causing the minified js file to be 50% larger.

Version 1.0.1-beta.2 (2015/12/04)
=================

  * support for ng-repeat-start/ng-repeat-end

Version 1.0.0-beta.1 (2015/12/04)
=================

  * bugfixes

Version 1.0.0-beta.0 (2015/12/04)
=================

### Major changes
  * elements are no longer absolutely positioned
  * no need for a mousewheel helper dummy element on MacOS (to preserve inertia scrolling)
  * no need for `top` updating directives - even better performance

Version 1.0.0-rc12 (2015/08/03)
=================

  * general improvements

Version 1.0.0-rc11 (2015/06/05)
=================

  * important bugfix

Version 1.0.0-rc10 (2015/06/03)
=================

  * You can now specify an attribute `vs-options="{latch: true}"` to enable latching mode - elements once rendered are not being removed when scrolled away (improves scrolling performance when the rendering of each element is time consuming)
  * bugfixes

Version 1.0.0-rc9 (2015/05/25)
=================

  * bugfixes

Version 1.0.0-rc8 (2015/05/25)
=================

  * support for `vs-scroll-parent="window"`
  * bugfixes

Version 1.0.0-rc7 (2015/05/25)
=================

bugfixes

Version 1.0.0-rc6 (2014/09/29)
=================

###FEATURES
  * added support for resizing items by changing the item size property and then `$broadcasting()`ing the ``vsRepeatTrigger`` event.

Version 1.0.0-rc5 (2014/08/01)
=================

###FEATURES
  * added an optional `vs-autoresize` attribute. You can use it without `vs-size-property` and without specifying element's size. The automatically computed element style will readjust upon window resize if the size is dependable on the viewport size.

Version 1.0.0-rc4 (2014/07/15)
=================

###FEATURES
  * added support for filters in the `ngRepeat` expression
  * added support for variable items sizes (if they are known up front)

Version 1.0.0-rc3 (2014/06/24)
=================
Fixed a bug on Safari and optimized layout recalculations

Version 1.0.0-rc2 (2014/06/09)
=================
Semver compatible version name

Version 1.0.0rc1 (2014/06/02)
=================

###FEATURES
- added travis-ci integration
- added unit tests
- added minified version of the library
- fixed some minor bugs
- updated the [DEMO](http://kamilkp.github.io/angular-vs-repeat/) that now also address the performance benefits of using the library

Version 0.3 (2014/05/30)
=================

###FEATURES
- support for automatic computation of repeated elements' `height`/`width`

###BREAKING CHANGE:
- specifying the single element's size as a value of the `vs-repeat` attribute is no longer required. If it isn't provided the directive will calculate it automatically (once, the first time it renders at least one element). If for some reason you want to override that size, you can still provide is as a value of the `vs-repeat` attribute.

Version 0.2 (2014/05/25)
=================

###FEATURES
- support for horizontally stacked elements (a `vs-horizontal` attribute must be placed on the element)
- no need to use jQuery, `vsRepeat` can now work with angular (jqLite) only
- support for before/after offsets (in pixels) via `vs-offset-before` and `vs-offset-after`
- full support for nested `vsRepeat`s and scrolling in both directions even within the same container

###BREAKING CHANGE:
- `vs-top-offset` attribute is now called `vs-offset-before`

Version 0.1
=================

###FEATURES
- support for vertically stacked elements


test/spec.js
/* global setInterval, console */

(() => {
  function stall() {
    setInterval(() => console.log(1), 1000);
    this.timeout(99999999999);
  }

  function getArray(size){
    const res = [];
    for (let i = 0; i < size; i++){
      res.push({
        value: i,
        size: 110,
      });
    }

    return res;
  }

  const getElements = $element => $element[0].querySelectorAll('.value');

  function getValues(elems){
    return Array.prototype.map.call(elems, function(elem){
      return +elem.innerHTML.trim();
    });
  }

  var animationFrame = 1000 / 60;

  describe('vs-repeat', function(){
    // this.timeout(1000000);
    var $compile,
      $element,
      $scope,
      $timeout;

    beforeEach(angular.mock.module('vs-repeat'));
    beforeEach(inject(function($injector, $rootScope){
      $compile = $injector.get('$compile');
      $timeout = $injector.get('$timeout');
      $element = null;
      $scope = $rootScope.$new();
      angular.element(document.head).append([
        `<style>
         	.container{
         		max-height: 200px;
         		max-width: 200px;
         		overflow: auto;
         		background: hsla(60, 100%, 50%, 0.3);
         	}
         	.scroll{
         		max-height: 100px;
         		max-width: 200px;
         		overflow: auto;
         		background: hsla(60, 100%, 50%, 0.3);
         	}
         	.container.horizontal{
            height: 200px;
            display: flex;
            flex-flow: row nowrap;
         	}
         	.item{
             width: 100%;
             height: 20px;
         	}
         	.horizontal .item{
            display: inline-block;
            width: auto;
         		min-width: 20px;
         		height: 100%;
         	}
        </style>`,
      ].join(''));
    }));
    afterEach(function(){
      if ($element)
        $element.remove();
      $scope.$destroy();
    });
    describe('automatic size', function(){
      it('should render only a few elements', function(done) {
        $element = $compile([
          '<div vs-repeat class="container">',
          '	<div ng-repeat="foo in bar" class="item">',
          '		<span class="value">{{foo.value}}</span>',
          '	</div>',
          '</div>',
        ].join(''))($scope);
        angular.element(document.body).append($element);
        $scope.bar = getArray(100);
        $scope.$digest();

        var elems, values1, values2, count;

        elems = getElements($element);
        values1 = getValues(elems);
        count = elems.length;
        expect(count).to.be.greaterThan(0);
        expect(count).to.be.lessThan(20);

        $element[0].scrollTop += 200;
        $element[0].scrollLeft += 200;

        $element.triggerHandler('scroll');

        elems = getElements($element);
        values2 = getValues(elems);
        expect(Math.min.apply(Math, values1)).to.be.lessThan(Math.min.apply(Math, values2));

        count = elems.length;
        expect(count).to.be.greaterThan(0);
        expect(count).to.be.lessThan(20);

        done();
      });

      it('should support rendering initially hidden element', function(done){
        $element = $compile([
          '<div ng-show="showFlag">',
          '	<div vs-repeat class="container">',
          '		<div ng-repeat="foo in bar" class="item">',
          '			<span class="value">{{foo.value}}</span>',
          '		</div>',
          '	</div>',
          '</div>',
        ].join(''))($scope);
        angular.element(document.body).append($element);
        $scope.bar = getArray(100);
        $scope.showFlag = false;
        $scope.$digest();

        setTimeout(function(){
          var elems = getElements($element);
          expect(elems.length).to.be.greaterThan(0);
          expect(elems.length).to.be.lessThan(4);

          $scope.showFlag = true;
          $scope.$digest();
          setTimeout(function(){
            elems = getElements($element);
            expect(elems.length).to.be.greaterThan(3);
            done();
          }, animationFrame);
        }, animationFrame);
      });
    });

    it('should support horizontal stacking of elements', function(done){
      $element = $compile([
        '<div vs-repeat="{horizontal: true}" class="container horizontal">',
        '	<div ng-repeat="foo in bar" class="item">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();

      var elems, values1, values2, count;

      elems = getElements($element);
      values1 = getValues(elems);
      count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      $element[0].scrollTop += 200;
      $element[0].scrollLeft += 200;

      $element.triggerHandler('scroll');

      elems = getElements($element);
      values2 = getValues(elems);
      expect(Math.min.apply(Math, values1)).to.be.lessThan(Math.min.apply(Math, values2));

      count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      done();
    });

    it('should support manually provided unique element size', function(done){
      $element = $compile([
        '<div vs-repeat="{size: 150}" class="container">',
        '	<div ng-repeat="foo in bar" class="item" ng-style="{height: \'150px\'}">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();

      setTimeout(() => {
        var elems = getElements($element);
        expect(elems.length).to.equal(2);
        done();
      });
    });

    it('should support manually provided size per element', function(done){
      $element = $compile([
        '<div vs-repeat="{size: \'foo.size\'}" class="container">',
        '	<div ng-repeat="foo in bar" class="item">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();
      $scope.$broadcast('vsRepeatTrigger');
      $scope.$digest();

      var elems = getElements($element);
      expect(elems.length).to.be.greaterThan(1);
      expect(elems.length).to.be.lessThan(5);
      done();
    });

    it('should support angular expression as vs-size attribute value', function(done){
      $element = $compile([
        '<div vs-repeat="{size: \'2*foo.size - 10\'}" class="container">',
        '	<div ng-repeat="foo in bar" class="item" ng-style="{ height: (2*foo.size - 10) + \'px\' }">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();
      $scope.$broadcast('vsRepeatTrigger');
      $scope.$digest();

      var elems = getElements($element);
      expect(elems.length).to.be.greaterThan(0);
      expect(elems.length).to.be.lessThan(3);
      done();
    });

    it('should support latching mode', function(done){
      $element = $compile([
        '<div vs-repeat="{size: \'foo.size\', latch: true}" class="container">',
        '	<div ng-repeat="foo in bar" class="item" ng-style="{ height: foo.size + \'px\' }">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();
      $scope.$broadcast('vsRepeatTrigger');
      $scope.$digest();

      var elems = getElements($element);
      expect(elems.length).to.be.greaterThan(1);
      expect(elems.length).to.be.lessThan(5);

      $element[0].scrollTop = 3000;
      $element.triggerHandler('scroll');
      expect(getElements($element).length).to.be.greaterThan(20);

      done();
    });

    it('should support changing manually provided size per element', function(done){
      $element = $compile([
        '<div vs-repeat="{size: \'foo.size\'}" class="container">',
        '	<div ng-repeat="foo in bar" class="item" ng-style="{ height: foo.size + \'px\' }">',
        '		<span class="value">{{foo.value}}</span>',
        '	</div>',
        '</div>',
      ].join(''))($scope);
      angular.element(document.body).append($element);
      $scope.bar = getArray(100);
      $scope.$digest();
      $scope.$broadcast('vsRepeatTrigger');
      $scope.$digest();

      var elems = getElements($element);
      expect(elems.length).to.be.greaterThan(1);
      expect(elems.length).to.be.lessThan(5);

      // Change first item size and trigger refresh
      var newSize = 123;
      $scope.bar[0].size = newSize;
      $scope.$digest();
      $scope.$broadcast('vsRepeatTrigger');
      $scope.$digest();

      // Second item should have shifted
      elems = getElements($element);
      expect(elems.length).to.be.greaterThan(1);
      expect(elems.length).to.be.lessThan(5);
      done();
    });

    it('should support non-default scrollable container', function(done){
      $element = angular.element([
        '<div class="scroll">',
        '	<div vs-repeat="{scrollParent: \'.scroll\'}">',
        '		<div ng-repeat="foo in bar" class="item">',
        '			<span class="value">{{foo.value}}</span>',
        '		</div>',
        '	</div>',
        '</div>',
      ].join(''));
      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      var elems, values1, values2, count;

      elems = getElements($element);
      values1 = getValues(elems);
      count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      $element[0].scrollTop += 200;
      $element[0].scrollLeft += 200;

      $element.triggerHandler('scroll');

      elems = getElements($element);
      values2 = getValues(elems);
      expect(Math.min.apply(Math, values1)).to.be.lessThan(Math.min.apply(Math, values2));

      count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      done();
    });

    it ('should support ngRepeater container selector', function (done) {
      $element = angular.element([
        '	<div vs-repeat vs-repeat-container=".container">',
        '	    <div class="container">',
        '		    <div ng-repeat="foo in bar" class="item">',
        '			    <span class="value">{{foo.value}}</span>',
        '		    </div>',
        '	    </div>',
        '	</div>',
      ].join(''));

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      var elems, count;

      elems = getElements($element);
      count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      done();
    });

    it('should work with ng-repeat-start', function (done) {
      $element = angular.element([
        '	<div vs-repeat vs-repeat-container=".container">',
        '	    <div class="container">',
        '		    <div ng-repeat-start="foo in bar" class="item">',
        '			    <span class="value">{{foo.value}}</span>',
        '		    </div>',
        '		    <div class="item">',
        '			    <span class="value">{{foo.value}}</span>',
        '		    </div>',
        '		    <div ng-repeat-end class="item">',
        '			    <span class="value">{{foo.value}}</span>',
        '		    </div>',
        '	    </div>',
        '	</div>',
      ].join(''));

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();
      $scope.$digest();

      const elems = getElements($element);
      const count = elems.length;
      expect(count).to.be.greaterThan(0);
      expect(count).to.be.lessThan(20);

      expect($element[0].innerText.startsWith('0\n0\n0')).to.be.true;

      $element.children()[0].scrollTop += 61;
      $element.triggerHandler('scroll');

      setTimeout(() => {
        expect($element[0].innerText.startsWith('1\n1\n1')).to.be.true;
        done();
      });
    });

    it('should execute function when scrolled to end offset', function (done) {
      $element = angular.element([
        '	<div vs-repeat="{size: 20, scrolledToEnd: \'updateCounter()\', scrolledToEndOffset: 20}" class="container">',
        '       <div ng-repeat="foo in bar" class="item" ng-style="{ height: \'20px\' }">',
        '		    <span class="value">{{foo.value}}</span>',
        '		</div>',
        '	</div>',
      ].join(''));

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      var counter = 0;
      $scope.updateCounter = function(){
        counter += 1;
      };

      $element[0].scrollTop = $element[0].scrollHeight * 0.6;
      $element.triggerHandler('scroll');
      $scope.$digest();

      expect(counter).to.equal(0);

      $element[0].scrollTop = $element[0].scrollHeight * 0.8;
      $element.triggerHandler('scroll');
      $scope.$digest();

      expect(counter).to.equal(1);

      done();
    });

    it ('should execute function when scrolled to beginning offset', function (done) {
      $element = angular.element([
        '	<div vs-repeat="{size: 20, scrolledToBeginning: \'updateCounter()\', scrolledToBeginningOffset: 20}" class="container">',
        '       <div ng-repeat="foo in bar" class="item" ng-style="{ height: \'20px\' }">',
        '		    <span class="value">{{foo.value}}</span>',
        '		</div>',
        '	</div>',
      ].join(''));

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      var counter = 0;
      $scope.updateCounter = function(){
        counter += 1;
      };

      $element[0].scrollTop = $element[0].scrollHeight * 0.8;
      $element.triggerHandler('scroll');
      $scope.$digest();

      expect(counter).to.equal(0);

      $element[0].scrollTop = 0;
      $element.triggerHandler('scroll');
      $scope.$digest();

      expect(counter).to.equal(1);

      done();
    });

    it('should properly calculate start and end index', function (done) {
      $element = angular.element(`
        <div vs-repeat="{size: 20}" class="container" style="height: 200px">
          <div ng-repeat="foo in bar" class="item">
            <span class="value">{{foo.value}}</span>
          </div>
        </div>
      `);

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      let elems = getElements($element);
      let count = elems.length;
      expect(count).to.equal(10);
      expect($element.scope().vsRepeat.startIndex).to.equal(0);
      expect($element.scope().vsRepeat.endIndex).to.equal(10);

      $element[0].scrollTop += 400;
      $element.triggerHandler('scroll');

      elems = getElements($element);
      count = elems.length;
      expect(count).to.equal(10);
      expect($element.scope().vsRepeat.startIndex).to.equal(20);
      expect($element.scope().vsRepeat.endIndex).to.equal(30);

      done();
    });

    it('should properly calculate start and end index with scroll margin', function (done) {
      $element = angular.element(`
        <div vs-repeat="{size: 20, scrollMargin: 40}" class="container" style="height: 200px">
          <div ng-repeat="foo in bar" class="item">
            <span class="value">{{foo.value}}</span>
          </div>
        </div>
      `);

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      let elems = getElements($element);
      let count = elems.length;
      expect(count).to.equal(12);
      expect($element.scope().vsRepeat.startIndex).to.equal(0);
      expect($element.scope().vsRepeat.endIndex).to.equal(12);

      $element[0].scrollTop += 400;
      $element.triggerHandler('scroll');

      elems = getElements($element);
      count = elems.length;
      expect(count).to.equal(14);
      expect($element.scope().vsRepeat.startIndex).to.equal(18);
      expect($element.scope().vsRepeat.endIndex).to.equal(32);

      done();
    });

    it('should properly calculate start and end index with element beyond', function (done) {
      $element = angular.element(`
        <div class="container" style="height: 200px;">
          <div vs-repeat="{size: 20, scrollParent: '.container'}" style="margin: 300px 0">
            <div ng-repeat="foo in bar" class="item">
              <span class="value">{{foo.value}}</span>
            </div>
          </div>
        </div>
      `);

      angular.element(document.body).append($element);
      $compile($element)($scope);
      $scope.bar = getArray(100);
      $scope.$digest();

      const $vsRepeat = $element.children().eq(0);

      let elems = getElements($element);
      let count = elems.length;
      expect(count).to.equal(0);
      expect($vsRepeat.scope().vsRepeat.startIndex).to.equal(0);
      expect($vsRepeat.scope().vsRepeat.endIndex).to.equal(0);

      $element[0].scrollTop = 500;
      $element.triggerHandler('scroll');

      elems = getElements($element);
      count = elems.length;
      expect(count).to.equal(10);
      expect($vsRepeat.scope().vsRepeat.startIndex).to.equal(10);
      expect($vsRepeat.scope().vsRepeat.endIndex).to.equal(20);

      $element[0].scrollTop = $element[0].scrollHeight + $element[0].clientHeight;
      $element.triggerHandler('scroll');

      elems = getElements($element);
      count = elems.length;
      expect(count).to.equal(0);
      expect($vsRepeat.scope().vsRepeat.startIndex).to.equal(100);
      expect($vsRepeat.scope().vsRepeat.endIndex).to.equal(100);

      done();
    });
  });
})();


.eslintrc.js
module.exports = {
  parser: 'babel-eslint',
  rules: {
    'prefer-const': ['error', { destructuring: 'all' }],
    'no-var': ['error'],
    'semi-spacing': ['error', { 'before': false, 'after': true }],
    'space-infix-ops': ['error'],
    'keyword-spacing': ['error'],
    'brace-style': ['error'],
    'no-else-return': ['error'],
    'semi': ['error'],
    'no-useless-concat': ['error'],
    'no-redeclare': ['error'],
    'no-undef': ['error'],
    'no-shadow': ['error'],
    'comma-dangle': ['error', 'always-multiline'],
    'indent': ['error', 2, {
      FunctionDeclaration: { parameters: 'first' },
      FunctionExpression: { parameters: 'first' },
      SwitchCase: 1,
      MemberExpression: 'off',
    }],
  },
  globals: {
    angular: false,
    document: false,
    window: false,
    expect: false,
    it: false,
    describe: false,
  }
};


karma.conf.js
module.exports = function(config) {
  config.set({
    frameworks: ['mocha', 'chai'],

    // list of files / patterns to load in the browser
    files: [
      'lib/angular.js',
      'dist/angular-vs-repeat.min.js',
      'lib/angular-mocks.js',
      'test/spec.js',
    ],

    // use dots reporter, as travis terminal does not support escaping sequences
    // possible values: 'dots', 'progress',
    // CLI --reporters progress
    reporters: ['progress'],

    // web server port
    // CLI --port 9876
    port: 9876,

    // enable / disable colors in the output (reporters and logs)
    // CLI --colors --no-colors
    colors: true,

    // level of logging
    // possible values: config.LOG_DISABLE || config.LOG_ERROR || config.LOG_WARN || config.LOG_INFO || config.LOG_DEBUG
    // CLI --log-level debug
    logLevel: config.LOG_INFO,

    // enable / disable watching file and executing tests whenever any file changes
    // CLI --auto-watch --no-auto-watch
    autoWatch: true,

    // Start these browsers, currently available:
    // - Chrome
    // - ChromeCanary
    // - Firefox
    // - Opera
    // - Safari (only Mac)
    // - PhantomJS
    // - IE (only Windows)
    browsers: [
      'Chrome',
      //'Firefox',
      //'Opera',
      //'IE',
    ],

    // If browser does not capture in given timeout [ms], kill it
    // CLI --capture-timeout 5000
    captureTimeout: 300000,

    // Auto run tests on start (when browsers are captured) and exit
    // CLI --single-run --no-single-run
    singleRun: false,

    // report which specs are slower than 500ms
    // CLI --report-slower-than 500
    reportSlowerThan: 5000,

    plugins: [
      'karma-mocha',
      'karma-chai',
      'karma-chrome-launcher',
      'karma-firefox-launcher',
    ],
  });
};

gulpfile.babel.js
import gulp from 'gulp';
import babel from 'gulp-babel';
import uglify from 'gulp-uglify';
import rename from 'gulp-rename';
import clean from 'gulp-clean';
import { Server } from 'karma';

gulp.task('babel', () =>
  gulp.src('src/angular-vs-repeat.js')
		.pipe(babel())
		.pipe(gulp.dest('dist'))
);

gulp.task('min', ['babel'], () =>
  gulp.src('dist/angular-vs-repeat.js')
		.pipe(uglify())
		.pipe(rename('angular-vs-repeat.min.js'))
		.pipe(gulp.dest('dist'))
);

gulp.task('karma', ['build'], (done) => {
  new Server({
    configFile: __dirname + '/karma.conf.js',
    singleRun: true,
  }, done).start();
});

gulp.task('karma-travis', ['build'], (done) => {
  new Server({
    configFile: __dirname + '/karma.conf.js',
    singleRun: true,
    browsers: [
      'Firefox',
    ],
  }, done).start();
});

gulp.task('build', ['babel', 'min']);
gulp.task('test', ['build', 'karma']);
gulp.task('travis', ['build', 'karma-travis']);


Gruntfile.js
module.exports = function(grunt){
	grunt.loadNpmTasks('grunt-contrib-jshint');
	grunt.loadNpmTasks('grunt-contrib-uglify');
	grunt.loadNpmTasks('grunt-karma');

	grunt.initConfig({
		uglify: {
			target: {
				files: {
					'src/angular-vs-repeat.min.js': ['src/angular-vs-repeat.js']
				},
				options: {
					banner: [
						'//',
						'// Copyright Kamil Pękala http://github.com/kamilkp',
						'// Angular Virtual Scroll Repeat v1.1.8 2017/09/15',
						'//',
						''
					].join('\n')
				}
			}
		},
		jshint: {
			all: [
				'src/angular-vs-repeat.js',
				'Gruntfile.js',
				'test/spec.js',
				'karma.conf.js'
			]
		},
		karma: {
			unit: {
				configFile: 'karma.conf.js',
				singleRun: true,
			},
			travis: {
				configFile: 'karma.conf.js',
				singleRun: true,
				browsers: [
					'Firefox'
				]
			}
		}
	});

	grunt.registerTask('min', 'Minify javascript source code', 'uglify');
	grunt.registerTask('test', 'Run unit tests', ['jshint', 'min', 'karma:unit']);
	grunt.registerTask('default', ['test']);
	grunt.registerTask('travis', ['jshint', 'min', 'karma:travis']);
};
